tests/drivers/can/api_additional/src/canfd.c

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/drivers/can.h>
+#include <zephyr/ztest.h>
+
+#include "common.h"
+
+/**
+ * @brief Send a CAN test frame with asserts.
+ *
+ * This function will block until the frame is queued or a test timeout
+ * occurs.
+ *
+ * @param dev      Pointer to the device structure for the driver instance.
+ * @param frame    Pointer to the CAN frame to send.
+ * @param callback Transmit callback function.
+ */
+static void send_test_frame(const struct device *dev, const struct can_frame *frame,
+				   can_tx_callback_t callback)
+{
+	int err;
+
+	err = can_send(dev, frame, TEST_SEND_TIMEOUT, callback, (void *)frame);
+	zassert_equal(err, 0, "failed to send frame (err %d)", err);
+}
+
+/* Additional tests */
+
+/**
+ * @brief Callback 1 invoked when CAN is stopped during transmission.
+ *
+ * See @a can_tx_callback_t() for argument description.
+ */
+static void tx_std_callback_netdown_1(const struct device *dev, int error, void *user_data)
+{
+	const struct can_frame *frame = user_data;
+
+	if ((error == 0) || (error == -ENETDOWN)) {
+		/* Frame may be sent before CAN is stopped */
+		k_sem_give(&tx_callback_sem);
+	}
+
+	zassert_equal(dev, can_dev, "CAN device does not match");
+	zassert_equal(frame->id, TEST_CAN_STD_ID_1, "ID does not match");
```

Comment:
Failed assert in callback may cause Zephyr fatal error, is this verified against this ?

Reply:
No it's not. So the test will fail anyway :smile: